doc/fluid/design/imperative/imperative.md

```diff
@@ -0,0 +1,99 @@
+# Imperative Programming
+
+This document will discuss changes needed in Paddle platform to support
+imperative programming.  
+
+Current users write Paddle provides a Python API that users typically 
+write their training/inference programs in.  Paddle will convert this 
+computation model into a ProgramDesc, which is executed in the C++ runtime.
+
+With this flexibility, and no clear guidelines on how to recommended usages
+of the Python API, a Paddle program could contain a mix of pure Python API
+and Paddle Fluid API.  In addition, the deep NN logic is combined with the
+logic to run it, which could further complicate the code.
+
+In addition, with the mix of Python and Paddle API, the ProgramDesc becomes
+impossible to be converted by Transpiler.
+
+## Changes to support Imperative
+
+Please refer to [Paddle API v4 Proposal](https://github.com/PaddlePaddle/Paddle/issues/9912)
+
+### Startup Program
+
+The Fluid startup programs are meant to be ran once before training starts.
+Its responsible for initilization tasks, including initializing variables and
+parameters.
+
+However since startup programs are created by the users in Python, and 
+execution initilized by the user in Python, this produces an issue for 
+imperative.  Users will need to know when to execute the startup program,
+and transpiler will have no clue how or when to execute it.
+
+##### Proposed Solution:
+
+With imperative programming, users may not need to directly interact with a
+startup program.  To support backwards compatibility, we will still indirectly
+create a startup and main program, however, before the program is executed, 
+we will merge the startup program with the main program and create a new 
+ProgramDesc.
+
+There is one challenge we will need to address however:
+- Since the startup and main programs are independent from each other, they
+may create block id and variable names that may conflict with each other
+when we merge the programs together.  We must take special care when 
+re-indexing the block ids and merging variables.
+
+### Training Loop
+
+With imperative, the training loop will be added to the main program as an 
+operator.  Since a majority of the use cases involves reading in training data
+and feeding the data to the network to be trained, we chose to create an 
+iterate method on reader to act as the training loop.
```

Comment:
What is the advantage of writing a specialized `ReaderOp` over writing a simple `IteratorOp`
```python
with Iterator(max_iter=1000):
    x, y = reader()
    ...
```

Reply:
Iterator op would introduce a general purpose control flow operator like While or IFElse, which are control flow operators that can be used for training purposes since they support back prop.

The ReaderOp iterator is specifically used for training loop (which should not be used for back prop).  If we introduce a general purpose op to iterate, we feel it would be confusing for users, since it does not behave like the other control flow ops.

Reply:
Sounds good. Thanks for the explanation.